src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs

```diff
@@ -0,0 +1,300 @@
+        /// </summary>
+        /// <param name="type"></param>
+        /// <returns></returns>
+        public static ReadOnlySymbolTable NewFromRecord(RecordType type)
+        {
+            return new SymbolTableOverRecordType(type ?? RecordType.Empty());
+        }
+
+        public static ReadOnlySymbolTable Compose(params ReadOnlySymbolTable[] tables)
+        {
+            // SymbolTableResolver walks Walks parents.
+            return new ComposedReadOnlySymbolTable(new SymbolTableEnumerator(tables));
+        }
+
+        // Helper to create a ReadOnly symbol table around a set of core functions.
+        // Important that this is readonly so that it can be safely shared across engines. 
+        internal static ReadOnlySymbolTable NewDefault(IEnumerable<TexlFunction> coreFunctions)
+        {
+            var s = new SymbolTable
+            {
+                EnumStoreBuilder = new EnumStoreBuilder(),
+                DebugName = $"BuiltinFunctions ({coreFunctions.Count()})"
+            };
+
+            foreach (var func in coreFunctions)
+            {
+                s.AddFunction(func); // will also add enum. 
+            }
+
+            return s;
+        }
+
+        private protected readonly Dictionary<string, NameLookupInfo> _variables = new Dictionary<string, NameLookupInfo>();
+
+        private protected readonly List<TexlFunction> _functions = new List<TexlFunction>();
+
+        // Which enums are available. 
+        // These do not compose - only bottom one wins. 
+        private protected EnumStoreBuilder _enumStoreBuilder;
+        private EnumSymbol[] _enumSymbolCache;
+
+        private EnumSymbol[] GetEnumSymbolSnapshot
+        {
+            get
+            {
+                // The caller may add to the builder after we've assigned. 
+                // So delay snapshot until we actually need to read it. 
+                if (_enumStoreBuilder == null)
+                {
+                    _enumSymbolCache = new EnumSymbol[] { };
+                }
+
+                if (_enumSymbolCache == null)
+                {
+                    _enumSymbolCache = _enumStoreBuilder.Build().EnumSymbols.ToArray();
+                }
+
+                return _enumSymbolCache;
+            }
+        }
+
+        internal IEnumerable<TexlFunction> Functions => ((INameResolver)this).Functions;
+
+        IEnumerable<TexlFunction> INameResolver.Functions => _functions;
+
+        IExternalDocument INameResolver.Document => default;
+
+        IExternalEntityScope INameResolver.EntityScope => throw new NotImplementedException();
+
+        IExternalEntity INameResolver.CurrentEntity => default;
+
+        DName INameResolver.CurrentProperty => default;
+
+        DPath INameResolver.CurrentEntityPath => default;
+
+        bool INameResolver.SuggestUnqualifiedEnums => false;
+
+        internal readonly Dictionary<DName, IExternalEntity> _environmentSymbols = new Dictionary<DName, IExternalEntity>();
```

Comment:
small readability thought:
Can we move all the default/notimpl left overs to the end/put it in a  region to hide or something like that?

Reply:
yes!